Services/JWSProvider.php

```diff
@@ -0,0 +1,192 @@
+            );
+        }
+
+        $this->keyLoader           = $keyLoader;
+        $this->encryptionEngine    = $encryptionEngine;
+        $this->encryptionAlgorithm = $encryptionAlgorithm;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createSignedToken(array $payload)
+    {
+        $this->jws = new JWS(['alg' => $this->encryptionAlgorithm], $this->encryptionEngine);
+        $this->jws->setPayload($payload);
+        $this->sign();
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadSignature($token)
+    {
+        $this->jws = JWS::load($token, false, null, $this->encryptionEngine);
+        $this->verify();
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getToken()
+    {
+        if (self::SIGNED !== $this->status) {
+            return;
+        }
+
+        return $this->getJWS()->getTokenString();
+    }
+
+    /**
+     * @return array $payload
+     */
+    public function getPayload()
+    {
+        return $this->getJWS()->getPayload();
+    }
+
+    /**
+     * @return string
+     */
+    public function getStatus()
+    {
+        return $this->status;
+    }
+
+    /**
+     * @link \Namshi\JOSE\SimpleJWS::isExpired()
+     *
+     * @return bool
+     */
+    public function isExpired()
+    {
+        $payload = $this->getJWS()->getPayload();
+
+        if (isset($payload['exp']) && is_numeric($payload['exp'])) {
+            $now = new \DateTime('now');
+
+            return ($now->format('U') - $payload['exp']) > 0;
+        }
+
+        return false;
+    }
+
+    /**
+     * Signs the created JWS signature.
+     */
+    protected function sign()
+    {
+        $this->getJWS()->sign(
+            $this->keyLoader->loadKey('private'),
+            $this->keyLoader->getPassphrase()
+        );
+
+        if (true === $this->getJWS()->isSigned()) {
+            $this->status = self::SIGNED;
+        }
+    }
+
+    /**
+     * Verifies the loaded JWS signature.
+     */
+    protected function verify()
+    {
+        $verified = $this->getJWS()->verify(
+            $this->keyLoader->loadKey('public'),
+            $this->encryptionAlgorithm
+        );
+
+        if (true === $verified) {
+            $this->status = self::VERIFIED;
+        }
+    }
+
+    /**
+     * @return JWS
+     *
+     * @throws \LogicException If there is no JWS instance defined.
+     */
+    protected function getJWS()
+    {
+        if (null === $this->jws) {
+            throw new \LogicException('There is no JWS instance defined. Did you forget to call %s::createSignedToken() or %s::loadSignature()?', get_class($this));
+        }
+
+        return $this->jws;
+    }
+
+    /**
+     * @param string $encryptionEngine
+     * @param string $encryptionAlgorithm
+     *
+     * @return bool
+     */
+    protected function isAlgorithmSupportedForEngine($encryptionEngine, $encryptionAlgorithm)
```

Comment:
Also, I would make *everything* `private`, unless you actively have a reason that you need `protected`. This will make it much easier to keep BC later. If someone needs to override something, they can open an issue and we can change to protected (or, we may be able to tell them the "correct" way to accomplish what they want).